mongo/gridfs/download_stream.go

```diff
@@ -128,14 +128,10 @@ func (ds *DownloadStream) Close() error {
 	return nil
 }
 
-// SetReadDeadline sets the read deadline for this download stream.
-func (ds *DownloadStream) SetReadDeadline(t time.Time) error {
-	if ds.closed {
-		return ErrStreamClosed
-	}
-
-	ds.readDeadline = t
-	return nil
+// WithContext sets the context for the DownloadStream, allowing control over
+// the execution and behavior of operations associated with the stream.
+func (ds *DownloadStream) WithContext(ctx context.Context) {
```

Comment:
An alternative to this would be to have a constructor that accepts a context for initializing a download stream.

Reply:
Do we need this setter at all? The only ways to create a `DownloadStream` are using `OpenDownloadStream` or `OpenDownloadStreamByName`, which both accept a `Context` parameter as of this PR.

Reply:
The context set by WithContext is specific to the read operation, which is independent of constructing a DownloadStream. For example, this:

```go
ctx, cancel := context.WithTimeout(context.Background(), -1*time.Second)
mt.Cleanup(cancel)

ds, err := bucket.OpenDownloadStreamByName(ctx, fileName) // could time out finding a file, etc
assert.Nil(mt, err, "OpenDownloadStreamByName error: %v", err)

p := make([]byte, len(fileData))
_, err = ds.Read(p)
```

has a different intent than this:

```go
ds, err := bucket.OpenDownloadStreamByName(context.Background(), fileName) 
assert.Nil(mt, err, "OpenDownloadStreamByName error: %v", err)

ctx, cancel := context.WithTimeout(context.Background(), -1*time.Second)
mt.Cleanup(cancel)

ds.WithContext(ctx) // specifically trying to add a context when reading a file

p := make([]byte, len(fileData))
_, err = ds.Read(p)
```

Reply:
I see your point about how the `Context` in `OpenDownloadStream` is only used to get the file info, but is not used for the subsequent ops that read the file info from the database. I think it's worth considering how similar APIs behave.

The Go stdlib offers a few examples of how to apply timeouts to stream reader types that implement `io.Reader`. The patterns are similar for applying timeouts to stream writer types that implement `io.Writer`.

The Go `net.Conn` allows setting a read timeout via `SetReadDeadline` or `SetDeadline`.
```go
conn, _ := net.Dial(...)
conn.SetReadDeadline(time.Now().Add(15 * time.Second))

// Will time out in 15 seconds
io.ReadAll(conn)
```

The Go `http.Client` allows setting a timeout that applies to the entire lifetime of any request, including dialing, reading headers, and reading the body.
```go
client := &http.Client{
    Timeout: 15 * time.Second,
}
resp, err := client.Get(...)

// Will time out in 15 seconds.
io.ReadAll(http.Body)
```

Thoughts:
* Concerning using "read deadline" vs "context", all of the underlying APIs used by the GridFS code accept a `Context` (they're all just Go driver CRUD calls), so using a `Context` seems to be the best choice.
* I think accepting a `Context` in `OpenDownloadStream` that is not used for actually downloading the file is confusing and would surprise most users. I recommend using the `Context` passed to `OpenDownloadStream` (and `OpenDownloadStreamByName`) as the `Context` on a `DownloadStream`.
* If we want to allow users to _override_ the `Context` used when actually downloading the file, we can add a `SetContext` method to `DownloadStream`. However, it's not immediately clear if that is necessary, so I'd recommend omitting it for now.

Reply:
The suggestion of WithContext comes directly from the http packages [Request.WithContext](https://pkg.go.dev/net/http#Request.WithContext) API. Which uses a context set by this method in it's [io operations](https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/net/http/request.go;l=560). I am also open to retaining the existing API which, as you note, is the pattern used in net.Conn. I would argue the existing pattern (SetReadDeadline) is unnecessarily asymmetric as DownloadStream does not have a concept of Write.

---

>I think accepting a Context in OpenDownloadStream that is not used for actually downloading the file is confusing and would surprise most users. I recommend using the Context passed to OpenDownloadStream (and OpenDownloadStreamByName) as the Context on a DownloadStream.

This sounds like a good idea.

Reply:
Another issue with going the constructor route is that if we ever needed to add a read-specific context timeout, then undoing the constructor propagation of context would be a breaking change. 

For example, suppose a user is setting a context on the constructor to timeout the the find operation, i.e. the construction. And they have no intention of attempting to timeout the io read. We would be tempted on the Go Driver team to add a WithContext method to DownloadStream to accommodate this case. However, we couldn't simply revert the context associated with the constructor because that could break another user's logic that expects a timeout to be shared between construction and read. This could be an awkward situation. 

I think simply having something like SetReadDeadline is the correct approach to the Download/Upload Stream objects.

Reply:
The [GridFS API](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst#gridfs-api) section of the CSOT spec actually describes the required behavior of the timeout param, which is basically "use the constructor context":
> ... all methods in the GridFS Bucket API MUST support the timeoutMS option. For methods that create streams (e.g. `open_upload_stream`), the option MUST cap the lifetime of the entire stream. ... Methods that interact with a user-provided stream (e.g. `upload_from_stream`) MUST use `timeoutMS` as the timeout for the entire upload/download operation.

Concerning the comment
> we couldn't simply revert the context associated with the constructor because that could break another user's logic that expects a timeout to be shared between construction and read

If we use the Context passed into the constructor, adding a new `WithContext` method to a `DownloadStream` doesn't seem like it would create a breaking change in API behavior.

For example, consider downloading a file with a 30 second timeout:
```go
ctx, cancel := context.WithTimeout(context.Background(), 30 * time.Second)
defer cancel()
ds, _ := bucket.OpenDownloadStream(ctx, ...)
b, _ := io.ReadAll(ds)
```

Now consider opening a `DownloadStream` with a 30 second timeout, but reading the file document(s) with no timeout:
```go
ctx, cancel := context.WithTimeout(context.Background(), 30 * time.Second)
defer cancel()
ds, _ := bucket.OpenDownloadStream(ctx, ...)
ds.WithContext(context.Background())
b, _ := io.ReadAll(ds)
```

Is there an examples where those timeouts would conflict?

It's still not clear that there is a use case for having different timeout behavior for different underlying operations during a GridFS upload/download, so I still recommend omitting it.

Reply:
@matthewdale For this:

> we couldn't simply revert the context associated with the constructor because that could break another user's logic that expects a timeout to be shared between construction and read

 I agree that there wouldn't be a conflict if (1) we didn't revert the context on the constructor, and (2) (probably) the WithContext method returned a shallow copy of the DownloadStream. **Consider this resolved**.

However, the primary issue I have with this approach is that the context lifecycle begins at construction. Consider your first case:

```go
ctx, cancel := context.WithTimeout(context.Background(), 30 * time.Second)
defer cancel()
ds, _ := bucket.OpenDownloadStream(ctx, ...)  // (*)
```
In your proposal, the context timeout doesn't just apply to the construction of `(*)`.  It would also be inherited by the DownloadStream object. So once `ds` is instantiated, you have ~30 seconds to read the download stream:

```go
// continuing
time.Sleep(45 *time.Second)
_, err := io.ReadAll(ds)
fmt.Println(err == context. DeadlineExceeded) // This is true
```
Here is a Go Playground example: https://go.dev/play/p/4ZZElYRypLW

Context values [shouldn't be stored on structs](https://go.dev/blog/context-and-structs#:~:text=Contexts%20should%20not%20be%20stored,store%20it%20in%20another%20type.), and the documentation linked covers this exact case:

> The caller’s lifetime is intermingled with a shared context, and the context is scoped to the lifetime where the Worker is created.

The docs also note that the only reason we should do this is for backwards-compatibility, which is not our issue in 2.x.

Unfortunately, if we want to time out the read operation, we have to do this. However, we can do it more modularly than at instantiation. `WithContext` gives us more control over what precisely a timeout effects.
---
Notes: 

The http packages [NewRequestWithContext](https://pkg.go.dev/net/http#NewRequestWithContext) also notes this:

> For an outgoing client request, the context controls the entire lifetime of a request and its response: obtaining a connection, sending the request, and reading the response headers and body.

Reply:
Agreed that we're basically using an antipattern, as described almost exactly in the "Storing context in structs leads to confusion" section of that "Contexts and structs" article. However, it's not significantly clearer if we provide a context via `WithContext` on a `DownloadStream` (I'd actually argue it's more confusing). It seems like we're designing an API to work around two problems:
1. `io.Reader` and `io.Writer` don't include contexts, so they basically have to be side-loaded for types that implement those interfaces. See an interesting proposal for adding contexts to those interfaces [here](https://github.com/golang/go/issues/20280).
2. The methods like `OpenDownloadStream` aren't an atomic "download a file" operation, but are currently the only way to accomplish downloading a file in the API described in the GridFS spec. That creates an conflict between Go Context best practices and the GridFS spec.

There's not much we can do about (1). However, we could separate the upload/download API into different methods, one supporting timeout and one not. For example, keep the existing methods with timeouts that only affect the initial operations but not the returned `DownloadStream` (i.e. there is no way to time out `Read` calls):
```go
func (b *Bucket) OpenDownloadStream(ctx context.Context, fileID any) (*DownloadStream, error)
```
Then add additional methods for upload/download that apply the context to the entire operation.
```go
func (b *Bucket) Download(ctx context.Context, fileID any, dst io.Writer) error
```

That deviates from the spec, but conforms more closely to Go Context best practices.

Reply:
From an offline conversation: The least surprising behavior is to make the Context apply to all I/O operations for a `DownloadStream` or `UploadStream`, which also matches the GridFS spec. The `Download` method I proposed above was intended to provide context and is really out of scope of this ticket, so can be ignored.